File: sdk/android/tokenpay-id-sdk/src/main/java/space/tokenpay/id/internal/PkceUtils.kt
package space.tokenpay.id.internal

import android.util.Base64
import java.security.MessageDigest
import java.security.SecureRandom

/**
 * Generates PKCE S256 verifier + challenge pairs per RFC 7636.
 */
internal object PkceUtils {
    private val secureRandom = SecureRandom()

    data class Pair(val verifier: String, val challenge: String, val method: String = "S256")

    /**
     * Generate a PKCE pair. Verifier is 43-char URL-safe base64 of 32 random bytes,
     * challenge = BASE64URL(SHA256(verifier)).
     */
    fun generate(): Pair {
        val verifierBytes = ByteArray(32).also { secureRandom.nextBytes(it) }
        val verifier = Base64.encodeToString(
            verifierBytes,
            Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP,
        )
        val sha = MessageDigest.getInstance("SHA-256").digest(verifier.toByteArray(Charsets.US_ASCII))
        val challenge = Base64.encodeToString(
            sha,
            Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP,
        )
        return Pair(verifier = verifier, challenge = challenge)
    }

    /**
     * Generate a cryptographically-random state string for CSRF protection.
     */
    fun randomState(sizeBytes: Int = 24): String {
        val bytes = ByteArray(sizeBytes).also { secureRandom.nextBytes(it) }
        return Base64.encodeToString(bytes, Base64.URL_SAFE or Base64.NO_PADDING or Base64.NO_WRAP)
    }

    /**
     * Generate a random nonce for OIDC.
     */
    fun randomNonce(): String = randomState(16)
}

File: sdk/swift/Sources/TokenPayID/Internal/PkceUtils.swift
import Foundation
import CryptoKit

/// PKCE S256 helper per RFC 7636.
enum PkceUtils {
    struct Pair: Sendable {
        let verifier: String
        let challenge: String
        let method = "S256"
    }

    static func generate() -> Pair {
        var bytes = Data(count: 32)
        _ = bytes.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, 32, $0.baseAddress!) }
        let verifier = base64URLEncode(bytes)
        let digest = SHA256.hash(data: Data(verifier.utf8))
        let challenge = base64URLEncode(Data(digest))
        return .init(verifier: verifier, challenge: challenge)
    }

    static func randomState(bytes: Int = 24) -> String {
        var buf = Data(count: bytes)
        _ = buf.withUnsafeMutableBytes { SecRandomCopyBytes(kSecRandomDefault, bytes, $0.baseAddress!) }
        return base64URLEncode(buf)
    }

    static func randomNonce() -> String { randomState(bytes: 16) }

    static func base64URLEncode(_ data: Data) -> String {
        data.base64EncodedString()
            .replacingOccurrences(of: "+", with: "-")
            .replacingOccurrences(of: "/", with: "_")
            .replacingOccurrences(of: "=", with: "")
    }
}
